### PR TITLE
Extended configuration of OOM injection mode

### PR DESCRIPTION
**Pull request by gerashegalov**
This PR enables specifying OOM injection mode with additional config

such as
- num_ooms to inject
- skip=N to make triger first OOM on N+1st allocation
- type with the current default mix `CPU_OR_GPU`,  `CPU` (just host allocations), `GPU` (just device allocations)
- split=bool whether to inject SplitAndRetryOOM

Enables IT invocations such as :

```
PYSP_TEST_spark_rapids_memory_gpu_state_debug=STDERR \
TEST_PARALLEL=0 \
SPARK_HOME=~/dist/spark-3.3.0-bin-hadoop3 \
./integration_tests/run_pyspark_from_build.sh \
  -k array_exists --test_oom_injection_mode=always:type=CPU,num_ooms=1,skip=4,split=true
``` 

This PR requires is stacked on NVIDIA/spark-rapids-jni#1637

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 1]
```diff
@@ -2296,11 +2297,11 @@ class RapidsConf(conf: Map[String, String]) extends Logging {
 
   lazy val perTaskOverhead: Long = get(TASK_OVERHEAD_SIZE)
 
-  lazy val concurrentGpuTasks: Int = get(CONCURRENT_GPU_TASKS)
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2299]: Why is this deleted? There were also several others deleted. If they are not used I understand, but at the same time I would prefer that to be in a separate PR.

**Review by revans2 — COMMENTED** [review]
Generally it looks good.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala (hunk 1 above)
Comment (gerashegalov) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2299]: It is unused, but understand the point of a separate PR

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 2]
```diff
@@ -1378,12 +1379,12 @@ object RapidsConf {
 
   // INTERNAL TEST AND DEBUG CONFIGS
 
-  val TEST_RETRY_OOM_INJECTION_ENABLED = conf("spark.rapids.sql.test.injectRetryOOM")
+  val TEST_RETRY_OOM_INJECTION_MODE = conf("spark.rapids.sql.test.injectRetryOOM")
     .doc("Only to be used in tests. If enabled the retry iterator will inject a GpuRetryOOM " +
          "or CpuRetryOOM once per invocation.")
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:1384]: Document needs to be updated.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala [hunk 3]
```diff
@@ -583,12 +583,13 @@ object RmmRapidsRetryIterator extends Logging {
...
-            // ensure we have associated our thread with the running task, as
-            // `forceRetryOOM` requires a prior association.
-            RmmSpark.currentThreadIsDedicatedToTask(TaskContext.get().taskAttemptId())
-            RmmSpark.forceRetryOOM(RmmSpark.getCurrentThreadId)
+          config.foreach {
+            case rapidsConf if !injectedOOM =>
+              injectedOOM = true
+              // ensure we have associated our thread with the running task, as
+              // `forceRetryOOM` requires a prior association.
+              rapidsConf.testRetryOOMInjectionMode.inject(TaskContext.get().taskAttemptId())
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala:591]: I'm not crazy about calling into the JNI every time when the injection mode is "false".

Comment (gerashegalov) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala:591]: ok, will update

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 4]
```diff
@@ -2300,7 +2304,51 @@ class RapidsConf(conf: Map[String, String]) extends Logging {
...
+        val injectConfMap = injectConfStr.split(',').map(_.split('=')).collect {
+          case Array(k, v) => k -> v
+        }.toMap
+        val numOoms = injectConfMap.getOrElse("num_ooms", 1.toString)
+        val skipCount = injectConfMap.getOrElse("skip", 0.toString)
+        val oomFilterStr = injectConfMap
+          .getOrElse("type", OomInjectionType.CPU_OR_GPU.toString)
+          .toUpperCase()
+        val oomFilter = OomInjectionType.valueOf(oomFilterStr)
+        val withSplit = injectConfMap.getOrElse("split", false.toString)
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2341]: I thought we talked about not adding "split" to the config format, because random splits will likely cause tests to fail.  We can force splits in unit test code via the RmmSpark interface.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 5]
```diff
@@ -2847,3 +2895,10 @@ class RapidsConf(conf: Map[String, String]) extends Logging {
     conf.contains(key)
   }
 }
+
+case class OomInjectionConf(
+  numOoms: Int,
+  skipCount: Int,
+  withSplit: Boolean,
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2902]: I think we can remove this.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala [hunk 6]
```diff
@@ -583,12 +583,25 @@ object RmmRapidsRetryIterator extends Logging {
...
+              // ensure we have associated our thread with the running task, as
+              // `forceRetryOOM` requires a prior association.
+              RmmSpark.currentThreadIsDedicatedToTask(TaskContext.get().taskAttemptId())
+              val injectConf = rapidsConf.testRetryOOMInjectionMode
+              if (injectConf.withSplit) {
+                RmmSpark.forceSplitAndRetryOOM(RmmSpark.getCurrentThreadId,
+                          injectConf.numOoms,
+                          injectConf.oomInjectionFilter.ordinal,
+                          injectConf.skipCount)
+              } else {
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala:598]: I don't think we want to force split and retries here.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala (hunk 4 above)
Comment (gerashegalov) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2341]: The default is `false`. I do not think it hurts  to have flexibility to schedule it for specific  allocation instance N. It may even be deterministic enough in some single core/batch/task scenarios.

**Comment by gerashegalov** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala (hunk 4 above)
Comment (jbrennan333) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2341]: It's ok to keep if you want, but please add it to the comment block above, and the config documentation string - we should probable also say something like "SplitAndRetry ooms are not always handled - use with care"

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 7]
```diff
@@ -1378,12 +1379,15 @@ object RapidsConf {
 
   // INTERNAL TEST AND DEBUG CONFIGS
 
-  val TEST_RETRY_OOM_INJECTION_ENABLED = conf("spark.rapids.sql.test.injectRetryOOM")
-    .doc("Only to be used in tests. If enabled the retry iterator will inject a GpuRetryOOM " +
-         "or CpuRetryOOM once per invocation.")
+  val TEST_RETRY_OOM_INJECTION_MODE = conf("spark.rapids.sql.test.injectRetryOOM")
+    .doc("Only to be used in tests. If `true` the retry iterator will inject a GpuRetryOOM " +
+         "or CpuRetryOOM once per invocation. Furthermore an extended config " +
+         "`num_ooms=<int>,skip=<int>,type=CPU|GPU|CPU_OR_GPU` can be provided to specify" +
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:1385]: Add split.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 8]
```diff
@@ -2300,7 +2304,51 @@ class RapidsConf(conf: Map[String, String]) extends Logging {
...
+   *
+   * "type" maps to OomInjectionType to run count against oomCount and skipCount
+   * "num_ooms" maps to oomCount (default 1), the number of allocations resulting in an OOM
+   * "skip" maps to skipCount (default 0), the number of matching  allocations to skip before
+   * injecting an OOM at the skip+1st allocation.
+   *
+   * For backwards compatibility support existing binary configuration
+   *   "false", disabled, i.e. oomCount=0, skipCount=0, injectionType=None
+   *   "true" or anything else but "false"  yields the default
+   *      oomCount=1, skipCount=0, injectionType=CPU_OR_GPU
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2321]: Add split

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 9]
```diff
@@ -1378,12 +1379,17 @@ object RapidsConf {
...
-  val TEST_RETRY_OOM_INJECTION_ENABLED = conf("spark.rapids.sql.test.injectRetryOOM")
-    .doc("Only to be used in tests. If enabled the retry iterator will inject a GpuRetryOOM " +
-         "or CpuRetryOOM once per invocation.")
+  val TEST_RETRY_OOM_INJECTION_MODE = conf("spark.rapids.sql.test.injectRetryOOM")
+    .doc("Only to be used in tests. If `true` the retry iterator will inject a GpuRetryOOM " +
+         "or CpuRetryOOM once per invocation. Furthermore an extended config " +
+         "`num_ooms=<int>,skip=<int>,type=CPU|GPU|CPU_OR_GPU,split=<bool>` can be provided to specify" +
+         "the number of OOMs to generate; how many to skip before doing so; whether to " +
+         "filter by allocation events by host(CPU), device(GPU), or both (CPU_OR_GPU); " +
+         "whether to inject *SplitAndRetryOOM in stead of plain *RetryOOM exceptions." +
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:1388]:
```suggestion
         "whether to inject *SplitAndRetryOOM instead of plain *RetryOOM exceptions." +
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 10]
```diff
@@ -2300,7 +2306,51 @@ class RapidsConf(conf: Map[String, String]) extends Logging {
...
+  /**
+   * Convert a string value to the injection configuration OomInjection.
+   *
+   * The new format is a CSV in any order
+   *  "num_ooms=<integer>,skip=<integer>,type=<string value of OomInjectionType>"
+   *
+   * "type" maps to OomInjectionType to run count against oomCount and skipCount
+   * "num_ooms" maps to oomCount (default 1), the number of allocations resulting in an OOM
+   * "skip" maps to skipCount (default 0), the number of matching  allocations to skip before
+   * injecting an OOM at the skip+1st allocation.
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2318]: Need to add `split=<bool>` along with description.

**Review by jbrennan333 — COMMENTED** [review]
Looking good - just a few minor nits, and it looks like prebuild checks are failing.

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
@jbrennan333 sorry missed one of  your comments. PTAL again.

**Review by jbrennan333 — APPROVED** [review]

**Comment by jbrennan333** [comment]
build

**Comment by gerashegalov** [comment]
build
